src/analyses/sese_regions.cpp

```diff
@@ -0,0 +1,261 @@
+        get_innermost_loop(innermost_loop_ids, it) ==
+          get_innermost_loop(innermost_loop_ids, possible_exit))
+      {
+        // If there are several candidate region exit nodes, prefer the one with
+        // the least dominators, i.e. the closest to the region entrance.
+        if(
+          closest_exit_index == dominators.cfg.size() ||
+          dominators.cfg[closest_exit_index].dominators.size() >
+            possible_exit_dominators)
+        {
+          closest_exit_index = possible_exit_index;
+        }
+      }
+    }
+
+    if(closest_exit_index < dominators.cfg.size())
+    {
+      auto emplace_result =
+        sese_regions.emplace(it, dominators.cfg[closest_exit_index].PC);
+      INVARIANT(
+        emplace_result.second, "should only visit each region entry once");
+    }
+  }
+}
+
+static std::string trimmed_last_line(const std::string &input)
+{
+  auto rhs_trim_idx = input.size() - 1;
+  while(rhs_trim_idx > 0 && isspace(input[rhs_trim_idx]))
+    --rhs_trim_idx;
+
+  auto lhs_trim_idx = input.rfind('\n', rhs_trim_idx);
+  if(lhs_trim_idx == std::string::npos)
+    lhs_trim_idx = 0;
+
+  while(lhs_trim_idx < input.size() && isspace(input[lhs_trim_idx]))
+    ++lhs_trim_idx;
+
+  return input.substr(lhs_trim_idx, (rhs_trim_idx - lhs_trim_idx) + 1);
+}
+
+typedef std::
+  unordered_map<goto_programt::const_targett, std::size_t, const_target_hash>
+    program_relative_instruction_indicest;
+
+static std::string instruction_ordinals(
+  const goto_programt::const_targett &instruction,
+  const program_relative_instruction_indicest
+    &program_relative_instruction_indices)
+{
+  return "(" +
+         std::to_string(program_relative_instruction_indices.at(instruction)) +
+         ", " + std::to_string(instruction->location_number) + ")";
+}
+
+static std::string brief_instruction_string(
+  const goto_programt &goto_program,
+  const goto_programt::const_targett &instruction,
+  const namespacet &ns,
+  const program_relative_instruction_indicest
+    &program_relative_instruction_indices)
+{
+  std::ostringstream ostr;
+  goto_program.output_instruction(ns, "", ostr, *instruction);
+  return instruction_ordinals(
+           instruction, program_relative_instruction_indices) +
+         " " + trimmed_last_line(ostr.str());
+}
+
+void sese_region_analysist::output(
+  std::ostream &out,
+  const goto_programt &goto_program,
+  const namespacet &ns) const
+{
+  program_relative_instruction_indicest program_relative_instruction_indices;
+  std::size_t next_index = 0;
+  for(auto it = goto_program.instructions.begin();
+      it != goto_program.instructions.end();
+      ++it, ++next_index)
+  {
+    program_relative_instruction_indices.emplace(it, next_index);
+  }
+  out << "Function contains " << sese_regions.size()
+      << " single-entry, single-exit regions:\n";
+  for(const auto &entry_exit : sese_regions)
+  {
+    out << "Region starting at "
+        << brief_instruction_string(
```

Comment:
Not sure why printing the location number is not sufficient, but ok.
Maybe this style of printing should be made available in goto instruction.

Reply:
Permits a reliable regression test, not sensitive to method ordering!